### PR TITLE
CorretFilter2DLinebuffer

### PR DESCRIPTION
**Pull request by ngdxzy**
The line buffer logic was technically wrong. The line0, line1, line2 should be strictly aligned with the filtering kernel. Therefore, line0 should always be the row above the central row; line 1 is always the central row; line2 is always the row below the central row. The original code is super random. The 'count' was not reset.
The previous error makes the output not consistent with the same input image.

Signed-off-by: Zhenyu Xu <zxu3@clemson.edu>

### Describe the problem solved by the commit
The line buffer logic in filter2d was wrong. It causes the following errors:

1. The output is not consistent with the same input image as the count is not reset.
2. The numerical result is wrong as line0, line1, line2 is not always aligned with the filtering kernel. It can be found with a Sobel Y filter ([[-1, -2, -1], [0, 0, 0], [1, 2, 1]). The comparison of output is provided in the screenshots. The testing code is provided in notebooks/LinebufferBUGfix/TestSobelY.ipynb

### How is the problem solved?
1. Reset static variable 'count' in linebuffer.h
2. Make sure line0 always be the row above the central row; line 1 is always the central row; line2 is always the row below the central row. 

### Are there any risks associated to the change?

It is tested with the official demo of filter2d examples.

### Is there a documentation impact?

Yes, the original code isn't correct.

### Checklist

- [ x] I added a test to cover my changes
- [ x] Existing and new test pass
- [ x] I read and I accept the [CONTRIBUTING.md](https://github.com/AMDResearch/Riallto/blob/main/CONTRIBUTING.md) guidelines

### Please provide screenshots (if applicable)
Correct sobelY:
image
Wrong sobelY (original linebuffer.h):
image

### Related issues
This code cannot solve the problem completely as the framework calls the filter kernel 720 times for 1280*720 figure for example. However, in the first call, the second row is impossible to be loaded into the line buffer. Theoretically, the kernel should be called 721 times. The first call is just to load row 0. The output starts from the second call.
Solving this problem completely may require significant code modification and cause incompatibility with other codes. Therefore, I currently partially solved this problem by ignoring the first row. In detail: the row 0 of the output is meaningless; the row 1 of the output is the correct output of row 0 with a mirrored boundary; row n in the output is always th row (n - 1) of the correct output.

### CODE REVIEW

File: npu/lib/kernels/cpp/linebuffer.h [hunk 1]
```diff
@@ -7,31 +7,51 @@ struct linebuffer_t{
...
+// line 0 always previous row/line
+// line 1 always central row/line
+// line 2 always next row/line
 template<int TWidth>
 linebuffer_t linebuffer(uint8_t * input, uint32_t num_lines) {
     static uint8_t linebuffer[3][TWidth];
     static uint32_t count = 0;
 
+    //     This is techniquly wrong as in the first call, only row 0 is loaded
+    // while first output pixel requires the first 2 rows. It cannot be solved
```
Comment (STFleming) [inline npu/lib/kernels/cpp/linebuffer.h:19]: Please remove comments, PR info describes the fix well enough.

**Review by STFleming — CHANGES_REQUESTED** [review]
This is great, thanks for the contribution!
Just a few small things:
-  The plan is to keep the notebook folder for educational notebooks, not for tests. Would it be possible to add a pytest instead? For this bugfix it is probably okay to remove the test if you'd prefer.
- Please remove the comments from the linebuffer.h file, the PR description describes the fix well enough.

Thanks again!

**Comment by ngdxzy** [comment]
> This is great, thanks for the contribution! Just a few small things:
> 
> * The plan is to keep the notebook folder for educational notebooks, not for tests. Would it be possible to add a pytest instead? For this bugfix it is probably okay to remove the test if you'd prefer.
> 
> Thanks again!

Sure, I have updated the commit according to your suggestions. Thanks a lot!

**Comment by STFleming** [comment]
Brilliant, thanks again for the contribution!

**Pull request merged**